sequence/model/lstm.py

```diff
@@ -0,0 +1,108 @@
+import torch
+import torch.nn.functional as F
+
+from sequence.model.modular import Embedding
+
+from torch import nn, optim
+
+import pickle
+import numpy as np
+
+
+class LSTM(Embedding):
+    def __init__(self, vocabulary_size, embedding_dim=10, custom_embeddings=None):
+        self.lstm_layers = 1
+        self.latent_size = 20
+        super().__init__(vocabulary_size, embedding_dim, custom_embeddings)
+
+        self.lstm = nn.LSTM(
+            input_size=embedding_dim,
+            hidden_size=self.latent_size,
+            num_layers=self.lstm_layers,
+            # dropout=0.2,
+            batch_first=True,
+        )
+        self.fc = nn.Linear(self.latent_size, vocabulary_size)
+
+    def forward(self, x, prev_state):
+        emb = self.apply_emb(x)
+        padded_emb, lengths = nn.utils.rnn.pad_packed_sequence(
+            emb, padding_value=0, batch_first=True
+        )
+        output, state = self.lstm(padded_emb, prev_state)
+        fc = self.fc(output)
+
+        return torch.log_softmax(fc, -1), state
+
+    def init_state(self, batch):
+        # TODO use random number for init
+
+        return (
+            torch.zeros(self.lstm_layers, batch, self.latent_size),
+            torch.zeros(self.lstm_layers, batch, self.latent_size),
+        )
+
+
+def det_loss(
+    model: LSTM,
+    packed_padded: nn.utils.rnn.PackedSequence,
+    test_loss: bool = False,
+    scale_loss_by_lengths: bool = False,
+    max_len: int = 1,
+    state_h: tuple = None,
+    state_c: tuple = None,
+):
+    # b,l,v
+    y_hat, (state_h, state_c) = model(packed_padded, (state_h, state_c))
```

Comment:
Seems to me that if you wrap you model in a fuction `Fn(packed_padded) -> y_hat` you could reuse the existing `det_loss` function, right? You could initialize `state_h` and `state_c` in the model implicitly.